src/main/scala/org/apache/spark/shuffle/ucx/ShuffleTransport.scala

```diff
@@ -0,0 +1,126 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION. ALL RIGHTS RESERVED.
+* See file LICENSE for terms.
+*/
+package org.apache.spark.shuffle.ucx
+
+import java.nio.ByteBuffer
+
+/**
+ * Class that represents some block in memory with it's address, size.
+ *
+ * @param isHostMemory host or GPU memory
+ */
+case class MemoryBlock(address: Long, size: Long, isHostMemory: Boolean = true)
+
+/**
+ * Opaque object to describe remote memory block (address, rkey, etc.).
+ */
+trait Cookie
+
+/**
+ * Base class to indicate some blockId. It should be hashable and could be constructed on both ends.
+ * E.g. ShuffleBlockId(shuffleId: Int, mapId: Long, reduceId: Int)
+ */
+trait BlockId
+
+trait Block {
+  // Transport will call this method when it would need an actual block memory.
+  def getMemoryBlock: MemoryBlock
```

Comment:
This trait must have a `releaseMemoryBlock` function. 

In the `mmap` case we do need to call `munmap` in order to unmap that memory, and remove references to shuffle files, for the files to be available to be unlinked.

If this doesn't support `mmap`'ed files, that's fine, but we need to figure out an alternative for disk bound shuffle files in that case.

Reply:
So if you don't need that block any more you need whether to call `transport.unregister(blockId)` of transport.mutate(blockId)`. After it's safe to do unmap (with mutate it's safe after callback)

Reply:
One other way of doing this is we keep a map of file-to-MemoryBlock outside of the transport. This would allow us to unmap it after a `remove` call or a `mutate` call succeeds. Was that the intended behavior? If so, I think I agree that we don't need `releaseMemoryBlock`.

Reply:
Yes, it's redundant. And transport may use different protocols (bcopy, one-sided). It's not obvious for transport when to call this method. But if you call `mutate` or `unregister` it's clear for transport that no need in this block